finat/finiteelementbase.py

```diff
@@ -144,6 +146,233 @@ def point_evaluation(self, order, refcoords, entity=None):
+            dual_functional = dual_functional_w_derivs[0]
+            for j in range(len(dual_functional)):
+                # Ignore alpha, just extract point (x_j) and weights (q_j)
+                x_j, q_j, _ = dual_functional[j]
+
+                # Esure all weights have the same shape
+                if last_shape is not None:
+                    assert q_j.shape == last_shape
+                last_shape = q_j.shape
+
+                assert q_j.children == ()
+                assert q_j.free_indices == ()
+
+                # Get k - the columns of Q.
+                x_matches = [x_j.almost_equal(x_k) for x_k in x]
+                if any(x_matches):
+                    k = x_matches.index(True)
+                else:
+                    k = len(x)
+                    x.append(x_j)
+
+                # q_j may be tensor valued
+                it = numpy.nditer(q_j.array, flags=['multi_index'])
+                for q_j_entry in it:
+                    Q[(i, k) + it.multi_index] = q_j_entry
+
+        if can_construct:
+
+            #
+            # CONVERT Q TO gem.SparseLiteral
+            #
+
+            # Convert dictionary of keys to SparseLiteral
+            Q = gem.SparseLiteral(Q)  # FIXME: This fails to compile since Impero can't yet deal with a sparse tensor
+            # FIXME: Temporarily use a normal literal
+            Q = gem.Literal(Q.array.todense())
+
+            #
+            # CONVERT x TO gem.PointSet
+            #
+
+            # Convert list of points to equivalent PointSet
+            allpts = None
+            dim = None
+            for i in range(len(x)):
+                # For the future - can only have one UnknownPointSingleton in the
+                # pointset.
+                # if isinstance(x[i], UnknownPointSingleton):
+                #     assert len(x) == 1
+                #     x = x[0]
+                # and skip x = PointSet(allpts)
+                if dim is not None:
+                    assert x[i].dimension == dim
+                dim = x[i].dimension
+                if allpts is not None:
+                    allpts = numpy.concatenate((allpts, x[i].points), axis=0)
+                else:
+                    allpts = x[i].points
+            assert allpts.shape[1] == dim
+            x = PointSet(allpts)
+
+            #
+            # EVALUATE fn AT x
+            #
+            expr = fn(x)
+
+            #
+            # TENSOR CONTRACT Q WITH expr
+            #
+            expr_shape_indices = tuple(gem.Index(extent=ex) for ex in expr.shape)
+            assert Q.free_indices == ()
+            Q_shape_indices = tuple(gem.Index(extent=ex) for ex in Q.shape)
+            assert tuple(i.extent for i in Q_shape_indices[2:]) == tuple(i.extent for i in expr_shape_indices)
+            basis_indices = Q_shape_indices[:1]
+            dual_eval_is = gem.index_sum(Q[basis_indices + x.indices + expr_shape_indices] * expr[expr_shape_indices], x.indices+expr_shape_indices)
```

Comment:
If you need to construct tensor contractions, please also consider using `gem.optimise.make_product` or `gem.optimise.sum_factorise`; they might give you extra benefits.

Reply:
Replacing 
```python
dual_eval_is = gem.index_sum(Q[basis_indices + x.indices + expr_shape_indices] * expr[expr_shape_indices], x.indices+expr_shape_indices)
```
with
```python
dual_eval_is = gem.optimise.make_product((Q[basis_indices + x.indices + expr_shape_indices], expr[expr_shape_indices]), x.indices+expr_shape_indices)
```
change the produced code from
```C
#include <stdint.h>

void expression_kernel(double *__restrict__ A, double const *__restrict__ w_0)
{
  double const t0[6 * 6] = { 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0 };
  double const t1[6 * 3] = { 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.5, 0.5, 0.5, 0.0, 0.5, 0.5, 0.5, 0.0 };
  double t2[6];

  for (int32_t i = 0; i <= 5; ++i)
  {
    t2[i] = 0.0;
    for (int32_t i_0 = 0; i_0 <= 2; ++i_0)
      t2[i] = t2[i] + t1[3 * i + i_0] * w_0[i_0];
  }
  for (int32_t p0 = 0; p0 <= 5; ++p0)
    for (int32_t i_1 = 0; i_1 <= 5; ++i_1)
      A[p0] = A[p0] + t0[6 * p0 + i_1] * t2[i_1];
}
```
to
```C
#include <stdint.h>

void expression_kernel(double *__restrict__ A, double const *__restrict__ w_0)
{
  double const t0[6 * 6] = { 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0 };
  double const t1[6 * 3] = { 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.5, 0.5, 0.5, 0.0, 0.5, 0.5, 0.5, 0.0 };
  double t2[6];

  for (int32_t i = 0; i <= 5; ++i)
  {
    t2[i] = 0.0;
    for (int32_t i_0 = 0; i_0 <= 2; ++i_0)
      t2[i] = t2[i] + t1[3 * i + i_0] * w_0[i_0];
  }
  for (int32_t p0 = 0; p0 <= 5; ++p0)
    for (int32_t i_1 = 0; i_1 <= 5; ++i_1)
      A[p0] = A[p0] + t0[6 * p0 + i_1] * t2[i_1];
}
```
i.e. it makes no difference at all.

If I avoid wrapping the result back up with `dual_eval_is_w_shape = gem.ComponentTensor(dual_eval_is, basis_indices)` and instead output the `basis_indices` directly the generated code still doesn't change

```C
#include <stdint.h>

void expression_kernel(double *__restrict__ A, double const *__restrict__ w_0)
{
  double const t0[6 * 6] = { 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0 };
  double const t1[6 * 3] = { 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.5, 0.5, 0.5, 0.0, 0.5, 0.5, 0.5, 0.0 };
  double t2[6];

  for (int32_t i = 0; i <= 5; ++i)
  {
    t2[i] = 0.0;
    for (int32_t i_0 = 0; i_0 <= 2; ++i_0)
      t2[i] = t2[i] + t1[3 * i + i_0] * w_0[i_0];
  }
  for (int32_t p0 = 0; p0 <= 5; ++p0)
    for (int32_t i_1 = 0; i_1 <= 5; ++i_1)
      A[p0] = A[p0] + t0[6 * p0 + i_1] * t2[i_1];
}
```

Reply:
I think you misinterpreted miklos' comment as being a "this will definitely fix things"? In this particular case that we're looking at now, the contraction is not complicated enough (or does not provide the structure such that) to be simplified by these routines.

However, if this is meant to be generic code, the comment is "if you're doing tensor contractions then you should consider delivering (potentially optimally scheduled) contractions by utilising the existing gem.optimise infrastructure" I think.

Reply:
Oooh it looks like there's delta cancellation in here! I'll try that

Reply:
Aww it doesn't seem to kick in for some reason :/